modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/ComponentMessageProcessorBuilder.java

```diff
@@ -152,10 +155,21 @@ public ComponentMessageProcessorBuilder<M, P> setTerminationTimeout(long termina
     return this;
   }
 
-  protected ConfigurationProvider getConfigurationProvider() {
-    return parameters.values().stream()
-        .filter(v -> v instanceof ConfigurationProvider)
-        .map(v -> ((ConfigurationProvider) v)).findAny()
-        .orElse(configurationProvider);
+  protected ConfigurationProviderResolverWrapper getConfigurationProviderResolver() {
+    // Uses the configurationProvider given to the builder if any, otherwise evaluates the parameters.
+    return configurationProvider != null ? new ConfigurationProviderResolverWrapper(configurationProvider)
+        : getConfigurationProviderResolver(parameters.get(CONFIG_ATTRIBUTE_NAME));
+  }
+
+  private ConfigurationProviderResolverWrapper getConfigurationProviderResolver(Object configRefParameter) {
+    if (configRefParameter instanceof ValueResolver) {
+      return new ConfigurationProviderResolverWrapper((ValueResolver<ConfigurationProvider>) configRefParameter);
+    }
+
+    if (configRefParameter instanceof ConfigurationProvider) {
+      return new ConfigurationProviderResolverWrapper((ConfigurationProvider) configRefParameter);
+    }
+
+    return null;
```

Comment:
I'm seeing if I can actually move this to an earlier stage so we already receive a ConfigProviderResolver and we can revert back to the original logic (only checking for a `ConfigurationProviderResolver` instead of `ConfigurationProvider`).

Maybe I can do that at the `ComponentConfigurationBuilder` which is the one that seems to be creating the resolvers.